core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryActivity.kt

```diff
@@ -0,0 +1,146 @@
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryListItem.HistoryItem
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.Filter
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.OnItemLongClick
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.ToggleShowHistoryFromAllBooks
+import org.kiwix.kiwixmobile.core.history.viewmodel.HistoryViewModel
+import org.kiwix.kiwixmobile.core.history.viewmodel.State
+import org.kiwix.kiwixmobile.core.history.viewmodel.State.Results
+import org.kiwix.kiwixmobile.core.history.viewmodel.State.SelectionResults
+import org.kiwix.kiwixmobile.core.utils.SimpleTextListener
+import java.util.ArrayList
+import javax.inject.Inject
+
+const val USER_CLEARED_HISTORY: String = "user_cleared_history"
+
+class HistoryActivity : OnItemClickListener, BaseActivity() {
+  private val activityComponent by lazy { coreActivityComponent }
+  private var selectedHistoryItems: List<ImageView> = ArrayList()
+  private val search = menu_history_search
+  private val deleteList: List<HistoryListItem> = ArrayList()
+  @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
+  private lateinit var recyclerView: RecyclerView
+  private val historyViewModel by lazy { viewModel<HistoryViewModel>(viewModelFactory) }
+  private val compositeDisposable = CompositeDisposable()
+
+  private val historyAdapter: HistoryAdapter2 by lazy {
+    HistoryAdapter2(
+      HistoryItemDelegate(this)
+    )
+  }
+
+  override fun injection(coreComponent: CoreComponent) {
+    activityComponent.inject(this)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_history)
+    setSupportActionBar(toolbar)
+
+    supportActionBar?.setDisplayHomeAsUpEnabled(true)
+    supportActionBar?.setTitle(string.history)
+
+    recycler_view.adapter = historyAdapter
+    recycler_view.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
+
+    history_switch.setOnCheckedChangeListener { _, isChecked ->
+      historyViewModel.actions.offer(ToggleShowHistoryFromAllBooks(isChecked))
+    }
+    history_switch.isChecked = !sharedPreferenceUtil.showHistoryCurrentBook
+
+    compositeDisposable.add(historyViewModel.effects.subscribe{it.invokeWith(this)})
+  }
+
+  override fun onDestroy() {
+    compositeDisposable.clear()
+    super.onDestroy()
+  }
+
+  override fun onCreateOptionsMenu(menu: Menu): Boolean {
+    menuInflater.inflate(R.menu.menu_history, menu)
+    val search = menu.findItem(id.menu_history_search).actionView as SearchView
+    search.queryHint = getString(string.search_history)
+    search.setOnQueryTextListener(SimpleTextListener {
+      historyViewModel.actions.offer(Filter(it))
+    })
+    historyViewModel.state.observe(this, Observer(::render))
+    return true
+  }
+
+  private fun render(state: State) {
+    when (state) {
```

Comment:
the coolest part of `when` is that when it is an expression it _has_ to be exhaustive or else there is a compile time error.
Because of that I usually turn `render` into an expression function.
It means you have to account for every state, very fun when adding new features, add a new state/action and the compiler tells you where you have to update

Reply:
That's actually very nice to have. Didn't know about that.